cloud/ec2_ami_copy.py

```diff
@@ -0,0 +1,224 @@
+
+    register_tags_if_any(module, ec2, tags, image_id)
+
+    module.exit_json(msg="AMI copy operation complete", image_id=image_id, state=img.state, changed=True)
+
+
+# register tags to the copied AMI in dest_region
+def register_tags_if_any(module, ec2, tags, image_id):
+    if tags:
+        try:
+            ec2.create_tags([image_id], tags)
+        except Exception as e:
+            module.fail_json(msg=str(e))
+
+
+# wait here until the image is copied (i.e. the state becomes available
+def wait_until_image_is_copied(module, ec2, wait_timeout, img, image_id, wait):
+    wait_timeout = time.time() + wait_timeout
+    while wait and wait_timeout > time.time() and (img is None or img.state != 'available'):
+        img = ec2.get_image(image_id)
+        time.sleep(3)
+    if wait and wait_timeout <= time.time():
+        # waiting took too long
+        module.fail_json(msg="timed out waiting for image to be copied")
+    return img
+
+
+# wait until the image is recognized.
+def wait_until_image_is_recognized(module, ec2, wait_timeout, image_id, wait):
+    for i in range(wait_timeout):
+        try:
+            return ec2.get_image(image_id)
+        except boto.exception.EC2ResponseError, e:
+            # This exception we expect initially right after registering the copy with EC2 API
+            if 'InvalidAMIID.NotFound' in e.error_code and wait:
+                time.sleep(1)
+            else:
+                # On any other exception we should fail
+                module.fail_json(
+                    msg="Error while trying to find the new image. Using wait=yes and/or a longer wait_timeout may help: " + str(
+                        e))
+    else:
+        module.fail_json(msg="timed out waiting for image to be recognized")
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        source_region=dict(required=True),
+        dest_region=dict(required=True),
+        source_image_id=dict(required=True),
+        name=dict(),
+        description=dict(default=""),
+        wait=dict(type="bool", default=False),
+        wait_timeout=dict(default=1200),
+        tags=dict()))
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    region, ec2_url, boto_params = get_aws_connection_info(module)
+    dest_region = module.params.get('dest_region')
+
+    if not module.params.get('dest_region'):
+        module.fail_json(msg='dest_region parameter is required')
+
+    try:
+        ec2 = connect_to_aws(boto.ec2, dest_region, **boto_params)
+    except boto.exception.NoAuthHandlerFound, e:
+        module.fail_json(msg=str(e))
+
+    if not module.params.get('source_region'):
+        module.fail_json(msg='source_region parameter is required')
+    if not module.params.get('source_image_id'):
+        module.fail_json(msg='source_image_id is required')
+
+    if module.params.get('tags'):
+        if not isinstance(module.params.get('tags'), dict):
```

Comment:
use `tags=dict(type=dict)` in the argument_spec to eliminate the need for this check